backend-lucene/src/main/java/org/hibernate/search/v6poc/backend/lucene/search/predicate/impl/BooleanJunctionPredicateBuilderImpl.java

```diff
@@ -85,47 +87,92 @@ private void addMinimumShouldMatchConstraint(int ignoreConstraintCeiling,
+		if ( shouldContributors == null ) {
+			shouldContributors = new ArrayList<>();
+		}
+		shouldContributors.add( contributor );
 	}
 
-	private void filter(Query luceneQuery) {
-		booleanQueryBuilder.add( luceneQuery, Occur.FILTER );
-		hasOnlyMustNot = false;
+	private void filter(SearchPredicateContributor<LuceneSearchPredicateContext, ? super LuceneSearchPredicateCollector> contributor) {
+		if ( filterContributors == null ) {
+			filterContributors = new ArrayList<>();
+		}
+		filterContributors.add( contributor );
 	}
 
 	@Override
-	protected Query buildQuery() {
-		if ( hasMustNot && hasOnlyMustNot ) {
+	protected Query buildQuery(LuceneSearchPredicateContext context) {
+		BooleanQuery.Builder booleanQueryBuilder = new BooleanQuery.Builder();
+
+		contributeQueries( context, booleanQueryBuilder, mustContributors, Occur.MUST );
+		contributeQueries( context, booleanQueryBuilder, mustNotContributors, Occur.MUST_NOT );
+		contributeQueries( context, booleanQueryBuilder, shouldContributors, Occur.SHOULD );
+		contributeQueries( context, booleanQueryBuilder, filterContributors, Occur.FILTER );
+
+		if ( isOnlyMustNot() ) {
 			booleanQueryBuilder.add( new MatchAllDocsQuery(), Occur.FILTER );
 		}
-		if ( minimumShouldMatchConstraints != null ) {
+
+		if ( minimumShouldMatchConstraints != null && shouldContributors != null ) {
 			int minimumShouldMatch;
 			Map.Entry<Integer, MinimumShouldMatchConstraint> entry =
-					minimumShouldMatchConstraints.lowerEntry( shouldClauseCount );
+					minimumShouldMatchConstraints.lowerEntry( shouldContributors.size() );
 			if ( entry != null ) {
-				minimumShouldMatch = entry.getValue().toMinimum( shouldClauseCount );
+				minimumShouldMatch = entry.getValue().toMinimum( shouldContributors.size() );
 			}
 			else {
-				minimumShouldMatch = shouldClauseCount;
+				minimumShouldMatch = shouldContributors.size();
 			}
 			booleanQueryBuilder.setMinimumNumberShouldMatch( minimumShouldMatch );
 		}
+
 		return booleanQueryBuilder.build();
 	}
 
+	private void contributeQueries(LuceneSearchPredicateContext context,
+			BooleanQuery.Builder booleanQueryBuilder,
+			List<SearchPredicateContributor<LuceneSearchPredicateContext, ? super LuceneSearchPredicateCollector>> contributors,
+			Occur occur) {
+		if ( contributors == null ) {
+			return;
+		}
+
+		for ( SearchPredicateContributor<LuceneSearchPredicateContext, ? super LuceneSearchPredicateCollector> contributor : contributors ) {
+			booleanQueryBuilder.add( getQueryFromContributor( context, contributor ), occur );
+		}
+	}
+
+	private boolean isOnlyMustNot() {
+		if ( mustNotContributors == null || mustNotContributors.isEmpty() ) {
+			return false;
+		}
+		if ( mustContributors != null && !mustContributors.isEmpty() ) {
+			return false;
+		}
+		if ( shouldContributors != null && !shouldContributors.isEmpty() ) {
+			return false;
+		}
+		if ( filterContributors != null && !filterContributors.isEmpty() ) {
+			return false;
+		}
+		return true;
```

Comment:
This looks more straightforward to my eyes, but YMMV:

```
return mustNotContributors != null && !mustNotContributors.isEmpty()
         && ( mustContributors == null || mustContributors.isEmpty() )
         && ( shouldContributors == null || shouldContributors.isEmpty() )
         && ( filterContributors == null || filterContributors.isEmpty() )
```

Reply:
Changed.